### ISSUE TITLE
Pin memory in subprocess

### ISSUE THREAD

**Issue by alexdremov** [pytorch/pytorch#130124]
### 🚀 The feature, motivation and pitch

Hey!

We are implementing custom data transforms pipeline. It involves subprocesses for parallel data processing. The pipeline works great, passing pytorch tensors with multipricessing queues via shared memory.

However, problems appear when trying to use `pin_memory()` in subprocesses. Unexpectedly, CUDA OOM start to appear. While I do not know the direct cause of such behavior, I suppose that it's due to the pinned memory caching allocator.

Is it possible to pin tensors without using torch's pinned memory caching allocator? Pinning memory in the main process greatly increases data processing time, which is suboptimal as pinning could be done in parallel subprocess. 

### Alternatives

Trivial solution is to pin in the main process, but such solution increases GPU wait time.

Also, generic torch DataLoader is not compatible with our processing pipeline due to data transforms complexity.

**Comment by andrewkho** [comment]
@alexdremov thanks for the issue! Can you share more about 
> generic torch DataLoader is not compatible with our processing pipeline due to data transforms complexity.
Any code pointers would be great. We're currently re-thinking this a little bit and hearing about why this doesn't work for you would help us guide our thinking.

For immediate workarounds:
In the torch.utils.data.DataLoader + multiprocessing, pin_memory will handle pinning in a background thread in the main process so that's an alternative you could explore.

**Comment by alexdremov** [comment]
> @alexdremov thanks for the issue! Can you share more about

Thanks for your reply! It's nice to know that this part of PyTorch is being refactored. 

While I cannot provide specific code examples, I can explain briefly why generic `DataLoader` does not fit our use case. This way, you may evaluate whether your planned architecture will eventually fit our requirements. For the context, we work with variable length samples. 

- First, we have dynamic batching. That is, there's no fixed batch size. Each batch is collected dynamically to ensure maximum GPU utilization. That is due to the memory consumption difference between "small batch of large samples" and "big batch of small samples". Currently, DataLoader can produce only fixed-sized batches. 

- Secondly, we have sophisticated data transforms. Our pipeline includes features calculations and augmentations. Not all transforms are equally CPU intensive. Therefore, some transforms are fine to be processed in a single worker, while others require dozens of CPU cores to provide acceptable throughput. Currently, DataLoader has `n` workers that do all the processing, without segregation between "light" transforms and "heavy" ones. 

- Some transforms must operate in a single process. For example, samples length sorting is a type of transform that is used to minimize number of paddings in batches. Clearly, it does not make sense to do length sorting in separate workers. That's type of functionality that is not currently available. 

- Profiling each pipeline step is complicated. To know how much time some parts of data processing take is essential for finding bottlenecks. I am not sure if it's currently possible. 

- Ordered mode. Some of our trainings require ordered data processing mode to ensure consistency and reproducibility. DataLoader's multi-worker mode has no such guarantees. 


As you see, list of drawbacks is quite large. Therefore, we implemented our own version of multiprocess data pipeline. In our solution, data pipeline is described in composable manner by YAML config. All data transforms starting from sample-level to the final batch are easily defined, with ability to specify how many processes to each part of data flow should be allocated. 

This way, we are able to retrieve batches from prefetch in no time. 

If you have any further questions, I am willing to answer them. You also can contact me directly by any contact in my profile. 

The only problem is that we cannot pin data in subprocesses, which increases GPU wait time. That's what the initial question is about. I will try to do so using pinning threads (similar to DataLoader manner), but this is still not the solution that I want, which is to retrieve pinned data from subprocesses.

**Comment by andrewkho** [comment]
Thanks @alexdremov for the very detailed list! One slight correction: current multiprocess dataloader does guarantee deterministic ordering by default. Beyond that, we don't currently have a pipelining/composability mechanism except for datapipes which is being deprecated in torchdata. Out of curiosity, when you mention different number of processes per transform step, are you communicating with multiprocess queues between the different transform stages? Have you noticed any overhead in these steps?

**Comment by alexdremov** [comment]
> Out of curiosity, when you mention different number of processes per transform step, are you communicating with multiprocess queues between the different transform stages? Have you noticed any overhead in these steps?

Yes, we use queues. These overheads are present, but they are not important as long as final prefetch queue is full. 

So, if data processing rate is higher than GPU consumption rate, then prefetch is full. In such case our data fetch time is `queue get time` + `pin memory`. The latter is dominating in terms of overhead. 

I have not explicitly measured how much time queue `get` takes, but we minimize the number of these synchronization by passing a batch of batches.
